### DISCUSSION TITLE
button#state=7 why not?

### DISCUSSION THREAD

**Discussion by BusinessClaes** [arendst/Tasmota#14270] [Q&A] [answered]
Just wondering why an I/O configured as button doesn’t give state=7 (CLEAR after long press). [Multipress and hold](https://tasmota.github.io/docs/Buttons-and-Switches/#multi-press-functions) states work as described i the docs, but I am really missing the CLEAR state.

Wouldn’t it be easy to implement as it doesn’t change any of the present behaviour/states of buttons?

I know that state=7 is available configuring the I/O as switch and Switchmode 11, but then there are other issues (only state=8 doublepress, no multipress, state=2 fires before state=8 upon doublepress, state=4 INC_DEC keeps firing during long press and I only need one eventnof state=3 which is notnimplemented in Switchmode 11)

I guess the background question is why a Switch with Switchmode 11 has so many different states compared to a Button. Wouldn’t it be more comprehensive to have only one or the other option, but with all the functionality/states? Or at least implement all the states in both options (as removing one of the options might create caos among already running rules all over the tasmota ecosystem).

But for now I would be really happy if someone with enough development skills could implement state=7 for buttons, that would make my day. Is this the right section for asking this? I didn’t want to put this as an issue, since I can’t call it a bug, just a feature request…

**Comment by sfromis** [comment, marked as answer]
Well, the idea is that if you want more than the simple setup with a Button, you can config it as a Switch input to have lots of flexibility. As things like hold and multipress can only be detected by waiting for the button to be released, it would not really make sense to have a clear action too. Basically you getting the detected button action is the clear too.

Looks like you're asking based on an idea about getting something to work, without asking about that "something".

**Reply by BusinessClaes** [reply]
Hmm… maybe I am getting it wrong, but I disagree. As far as I can see the event button#state=3 (HOLD) comes after the long-press surpasses the time defined in SetOption32, it doesn’t wait until the button is released. So the information about when the long-press is released gets lost.

Just to clarify, I agree the CLEAR is completely unnecessary after single and multipress, but I definitely think it is needed after a long-press.

And of course, I am trying to implement my “something” but I thought it was better to ask the generic question instead of bothering with my specific needs.
But here it comes :)
I have changed the Blackadder template of my Shelly Dimmer 2 I/O so I can use my momentary switch as a Button not as a Switch, since I want to have multipress for different actions along with hold and release for dimming control. Having the I/O defined as switch with Switchmode 11 allows only double-press which I’d rather not be restricted to. And I also prefer the behaviour of the button#state=3 firing only once instead of the switch#state=4 coming repeatedly during long-press. Having only one event upon long-press allows me to use the command Dimmer>/Dimmer< which results in a smooth dimming compared to hacky multiple Dimmer+/Dimmer- commands. But then I need the button#state=7 to execute the command Dimmer! to stop the dimming, and sadly the state=7 is not available.

**Reply by sfromis** [reply]
Clear "needed"? Certainly not, even if you may be able to envision a use case for that added complexity.

IMO, you may be trying to do "too much" with one button. I'm not really a fan of a button needing a user manual due to a pattern of presses and timing to take different actions.

**Reply by BusinessClaes** [reply]
Ok, you are probably right, I might be trying to put too many features with multipress in the same button… But I still think (independently of my particular case here) that knowing when a long-press is released qualifies as a basic feature for a button. Do you really don’t think it should be available without needing to go the Switchmode 11 way?

**Reply by sfromis** [reply]
To me, long-press release time, after expiry of the detection threshold, is far from a "basic feature". Of course "anything" may still be seen as useful to users wanting to make the button usage too complex ;-)
